src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs

```diff
@@ -164,9 +164,24 @@ internal ActiveHandlerTrackingEntry CreateHandlerEntry(string name)
 
                 void Configure(HttpMessageHandlerBuilder b)
                 {
-                    for (int i = 0; i < options.HttpMessageHandlerBuilderActions.Count; i++)
+                    if (options.MergedToHandlerBuilderActions)
                     {
-                        options.HttpMessageHandlerBuilderActions[i](b);
+                        for (int i = 0; i < options.HttpMessageHandlerBuilderActions.Count; i++)
+                        {
+                            options.HttpMessageHandlerBuilderActions[i](b);
+                        }
+                    }
+                    else
+                    {
+                        for (int i = 0; i < options.PrimaryHandlerActions.Count; i++)
+                        {
+                            options.PrimaryHandlerActions[i](b);
+                        }
+
+                        for (int i = 0; i < options.AdditionalHandlersActions.Count; i++)
+                        {
+                            options.AdditionalHandlersActions[i](b);
+                        }
```

Comment:
If you change PrimaryHandlerActions / AdditionalHandlersActions to be `List<T>` instead of `IReadOnlyList<T>`, then you can `foreach` these, instead.